### PR TITLE
[ModelBash][DSV3] Add TRTLLM DSV3 Router GEMM kernel (6% B1 Speedup)

### PR DESCRIPTION
**Pull request by robertgshaw2-redhat**
Port the optimized router GEMM kernel from sglang's sgl-kernel for DeepSeek V3 MoE models. This kernel is specifically optimized for small batch sizes (1-16 tokens) common in decode phase. The kernel is originally adapted from TRTLLM.
- original PR: https://github.com/sgl-project/sglang/pull/7627/changes

Key features:
- Computes output = mat_a @ mat_b.T for MoE routing
- Supports bfloat16 input with float32 or bfloat16 output (router logits uses fp32)
- Optimized for DSV3 dimensions: hidden_dim=7168, num_experts={256,384}
- Requires SM90+ (Hopper) GPUs and CUDA 12.0+
- Supports Programmatic Dependent Launch (PDL) via TRTLLM_ENABLE_PDL=1

Original kernel adapted from TensorRT-LLM's dsv3RouterGemm implementation.

5.5% E2E Speedup for Batch 1 Decode.

## Test Plan

```bash
eval:
	lm_eval \
		--model local-completions \
		--tasks gsm8k \
		--model_args "model={{MODEL}},base_url=http://localhost:{{PORT}}/v1/completions,num_concurrent=10,tokenized_requests=False" --limit 100
```
^ run with concurrency 10 to hit the low batch size

```bash
benchmark:
	vllm bench serve \
		--port {{PORT}} \
		--model {{MODEL}} \
		--dataset-name random \
		--input-len 2 \
		--output-len 100 \
		--max-concurrency 1 \
		--num-prompts 10 \
		--seed $(date +%s) \
		--temperature 0.0
```

## Test Result

- pr accuracy
```bash
local-completions ({'model': 'nvidia/DeepSeek-V3.1-NVFP4', 'base_url': 'http://localhost:8001/v1/completions', 'num_concurrent': 10, 'tokenized_requests': False}), gen_kwargs: ({}), limit: 100.0, num_fewshot: None, batch_size: 1
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  | 0.97|±  |0.0171|
|     |       |strict-match    |     5|exact_match|↑  | 0.97|±  |0.0171|
```

- main
```bash
============ Serving Benchmark Result ============
Successful requests:                     10        
Failed requests:                         0         
Maximum request concurrency:             1         
Benchmark duration (s):                  9.13      
Total input tokens:                      10        
Total generated tokens:                  1000      
Request throughput (req/s):              1.10      
Output token throughput (tok/s):         109.58    
Peak output token throughput (tok/s):    111.00    
Peak concurrent requests:                2.00      
Total token throughput (tok/s):          110.67    
---------------Time to First Token----------------
Mean TTFT (ms):                          21.32     
Median TTFT (ms):                        18.55     
P99 TTFT (ms):                           43.17     
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          9.00      
Median TPOT (ms):                        9.00      
P99 TPOT (ms):                           9.02      
---------------Inter-token Latency----------------
Mean ITL (ms):                           9.00      
Median ITL (ms):                         8.99      
P99 ITL (ms):                            9.34      
==================================================
```

- pr
```bash
============ Serving Benchmark Result ============
Successful requests:                     10        
Failed requests:                         0         
Maximum request concurrency:             1         
Benchmark duration (s):                  8.65      
Total input tokens:                      10        
Total generated tokens:                  1000      
Request throughput (req/s):              1.16      
Output token throughput (tok/s):         115.65    
Peak output token throughput (tok/s):    117.00    
Peak concurrent requests:                3.00      
Total token throughput (tok/s):          116.80    
---------------Time to First Token----------------
Mean TTFT (ms):                          22.68     
Median TTFT (ms):                        18.25     
P99 TTFT (ms):                           58.48     
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          8.50      
Median TPOT (ms):                        8.50      
P99 TPOT (ms):                           8.53      
---------------Inter-token Latency----------------
Mean ITL (ms):                           8.50      
Median ITL (ms):                         8.50      
P99 ITL (ms):                            8.88      
==================================================
```

image

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

**Comment by mgoin** [comment]
Can we use the router gemm interface already present in flashinfer? 
* DSV3 https://github.com/flashinfer-ai/flashinfer/pull/2019
* ML3 https://github.com/flashinfer-ai/flashinfer/pull/2323

**Review by LucasWilkinson — APPROVED** [review]
Nice! LGTM, thanks!

File: CMakeLists.txt [hunk 1]
```diff
@@ -1082,6 +1082,23 @@ if(VLLM_GPU_LANG STREQUAL "CUDA")
     message(STATUS "Not building Marlin MOE kernels as no compatible archs found"
                    " in CUDA target architectures")
   endif()
+
+  # DeepSeek V3 router GEMM kernel - requires SM90+
+  cuda_archs_loose_intersection(DSV3_ROUTER_GEMM_ARCHS "9.0a;10.0a" "${CUDA_ARCHS}")
```
Comment (mgoin) [inline CMakeLists.txt]: This isn't compatible with CUDA 13 and missing blackwell ultra, should be something like
```
  if(${CMAKE_CUDA_COMPILER_VERSION} VERSION_GREATER_EQUAL 13.0)
    cuda_archs_loose_intersection(DSV3_ROUTER_GEMM_ARCHS "9.0a;10.0f;11.0f" "${CUDA_ARCHS}")
  else()
    cuda_archs_loose_intersection(DSV3_ROUTER_GEMM_ARCHS "9.0a;10.0a;10.1a;10.3a" "${CUDA_ARCHS}")
  endif()
```

File: vllm/model_executor/models/deepseek_v2.py [hunk 2]
```diff
@@ -220,6 +221,48 @@ def forward(self, x):
...
+            quant_config=quant_config,
+            prefix=f"{prefix}.gate",
+        )
+
+        self.allow_dsv3_router_gemm = True
+
+    def _set_allow_dsv3_router_gemm(self) -> None:
+        self.allow_dsv3_router_gemm = (
+            current_platform.is_cuda()
+            and current_platform.has_device_capability((9, 0))
```
Comment (mgoin) [inline vllm/model_executor/models/deepseek_v2.py]: This should be `current_platform.is_device_capability(90) and current_platform.is_device_capability_family(100)` since we aren't supporting sm120

Comment (mgoin) [reply vllm/model_executor/models/deepseek_v2.py]: It also looks like you need to check against supported n_experts since I only see instantiations for 256 or 384 experts

File: csrc/moe/dsv3_router_gemm_utils.h [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+inline int getSMVersion() {
+  int device{-1};
+  cudaGetDevice(&device);
+  int sm_major = 0;
+  int sm_minor = 0;
+  cudaDeviceGetAttribute(&sm_major, cudaDevAttrComputeCapabilityMajor, device);
+  cudaDeviceGetAttribute(&sm_minor, cudaDevAttrComputeCapabilityMinor, device);
+  return sm_major * 10 + sm_minor;
+}
```
Comment (mgoin) [inline csrc/moe/dsv3_router_gemm_utils.h:31]: nit: we should already have a util for this

File: csrc/moe/dsv3_router_gemm_entry.cu [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+  TORCH_CHECK(num_tokens >= 1 && num_tokens <= 16,
+              "currently num_tokens must be less than or equal to 16 for "
+              "router_gemm");
+  TORCH_CHECK(mat_a.dtype() == at::kBFloat16, "mat_a must be bf16");
+  TORCH_CHECK(mat_b.dtype() == at::kBFloat16, "mat_b must be bf16");
+  TORCH_CHECK(output.dtype() == at::kFloat || output.dtype() == at::kBFloat16,
+              "output must be float32 or bf16");
+
+  auto const sm = getSMVersion();
+  TORCH_CHECK(sm >= 90, "required CUDA ARCH >= SM_90");
```
Comment (mgoin) [inline csrc/moe/dsv3_router_gemm_entry.cu]: Do you know if this would work on SM120? Better to be explicit if we don't know

**Review by mgoin — CHANGES_REQUESTED** [review]

File: csrc/moe/dsv3_router_gemm_utils.h (hunk 3 above)
Comment (robertgshaw2-redhat) [reply csrc/moe/dsv3_router_gemm_utils.h:31]: any pointer? sorry im not too familiar with the cuda code

**Review by mgoin — APPROVED** [review]
LGTM if eval and perf is good

File: csrc/moe/dsv3_router_gemm_utils.h (hunk 3 above)
Comment (robertgshaw2-redhat) [reply csrc/moe/dsv3_router_gemm_utils.h:31]: I looked into it. Seems like its tied into cutlass, simplified it a bit

**Review by mgoin — APPROVED** [review]
LGTM

**Comment by robertgshaw2-redhat** [comment]
@xinli-sw - sounds good. We can add it once FI once flashinfer hits 0.6.5

**Comment by robertgshaw2-redhat** [comment]
> no problem, it's a big project with a very wide hardware support. Stuff happens.

I have to say, I dont quite guy why this did not break on SM89 where we run a lot of tests.
